superdesk/users/async_service.py

```diff
@@ -0,0 +1,619 @@
+
+        items_locked_by_user = archive_service.get_from_mongo(req=None, lookup=query)
+
+        if items_locked_by_user and items_locked_by_user.count():
+            for item in items_locked_by_user:
+                # delete the item if nothing is saved so far
+                if item[VERSION] == 0 and item["state"] == "draft":
+                    get_resource_service("archive").delete(lookup={"_id": item["_id"]})
+                else:
+                    archive_service.update(item["_id"], doc_to_unlock, item)
+                    archive_autosave_service.delete(lookup={"_id": item["_id"]})
+
+    async def on_deleted(self, doc: UsersResourceModel):
+        """Overriding to add to activity stream and handle user clean up.
+
+        1. Authenticated Sessions
+        2. Locked Articles
+        3. Reset Password Tokens
+        """
+
+        add_activity(
+            ACTIVITY_UPDATE,
+            "disabled user {{user}}",
+            self.datasource,  # TODO-ASYNC: Confirm datasource change
+            user=doc.to_dict().get("display_name", doc.to_dict().get("username")),
+        )
+        await self.__clear_locked_items(str(doc.to_dict().get("_id")))
+        await self.__handle_status_changed(updates={"is_enabled": False, "is_active": False}, user=doc)
+
+    async def on_fetched(self, document):
+        for doc in document["_items"]:
+            await self.__update_user_defaults(doc)
+
+    async def on_fetched_item(self, doc: UsersResourceModel):
+        await self.__update_user_defaults(doc)
+
+    async def __update_user_defaults(self, doc: UsersResourceModel):
+        """Set default fields for users"""
+        user_dict = doc.to_dict()
+        user_dict.pop("password", None)
+        user_dict.setdefault("display_name", get_display_name(doc))
+        user_dict.setdefault("is_enabled", user_dict.get("is_active"))
+        user_dict.setdefault(SIGN_OFF, set_sign_off(doc))
+        user_dict["dateline_source"] = get_app_config("ORGANIZATION_NAME_ABBREVIATION")
+        doc = UsersResourceModel(**user_dict)
+
+    async def user_is_waiting_activation(self, doc: UsersResourceModel):
+        return doc.to_dict().get("needs_activation", False)
+
+    async def is_user_active(self, doc: UsersResourceModel):
+        return doc.to_dict().get("is_active", False)
+
+    async def get_role(self, user: UsersResourceModel):
+        if user:
+            role_id = user.to_dict().get("role", None)
+            if role_id:
+                return get_resource_service("roles").find_one(_id=role_id, req=None)
+        return None
+
+    async def set_privileges(self, user: UsersResourceModel, role):
+        user_dict = user.to_dict()
+        user_dict["active_privileges"] = get_privileges(user, role)
+        user = UsersResourceModel(**user_dict)
+
+    # TODO-ASYNC: Confirm on change to `get` function
+    async def get(self, req, lookup) -> ResourceCursorAsync[UsersResourceModel]:
```

Comment:
This entire function can be skipped in this async service, as it is in the context of an API Request, so it belongs in the REST API service (which we will do at a later stage).